File: minmaxheap/medianofstream/MedianOfStream.java
package medianofstream;
import java.util.PriorityQueue;

class MedianOfStream {
    private PriorityQueue<Integer> lowerHalfMaxHeap;  // Max heap to store the smaller half of the numbers
    private PriorityQueue<Integer> upperHalfMinHeap;  // Min heap to store the larger half of the numbers

    public MedianOfStream() {
        lowerHalfMaxHeap = new PriorityQueue<>((a, b) -> b - a);  // Max heap to store the lower half
        upperHalfMinHeap = new PriorityQueue<>();  // Min heap to store the upper half
    }

    // Method to insert a number into the stream
    public void insertNumber(int number) {
        // Balance the heaps by alternating insertion between them
        if (lowerHalfMaxHeap.size() == upperHalfMinHeap.size()) {
            lowerHalfMaxHeap.offer(number);
            upperHalfMinHeap.offer(lowerHalfMaxHeap.poll());
        } else {
            upperHalfMinHeap.offer(number);
            lowerHalfMaxHeap.offer(upperHalfMinHeap.poll());
        }
    }

    // Method to find the median of the numbers inserted so far
    public double findMedian() {
        // If both heaps are the same size, the median is the average of the two middle elements
        if (lowerHalfMaxHeap.size() == upperHalfMinHeap.size()) {
            return (lowerHalfMaxHeap.peek() + upperHalfMinHeap.peek()) / 2.0;
        } else {
            // Otherwise, the median is the top element of the heap with more elements (upperHalfMinHeap)
            return upperHalfMinHeap.peek();
        }
    }

    public static void main(String[] args) {
        MedianOfStream medianCalculator = new MedianOfStream();
        medianCalculator.insertNumber(3);
        medianCalculator.insertNumber(1);
        System.out.println(medianCalculator.findMedian());  // Output: 2.0
        medianCalculator.insertNumber(5);
        System.out.println(medianCalculator.findMedian());  // Output: 3.0
        medianCalculator.insertNumber(4);
        System.out.println(medianCalculator.findMedian());  // Output: 3.5
    }
}

File: minmaxheap/medianofstream/MedianOfStream.py
import heapq

class MedianOfStream:
    def __init__(self):
        self.small = []  # max heap
        self.large = []  # min heap

    def insertNum(self, num: int) -> None:
        if len(self.small) == len(self.large):
            heapq.heappush(self.large, -heapq.heappushpop(self.small, -num))
        else:
            heapq.heappush(self.small, -heapq.heappushpop(self.large, num))

    def findMedian(self) -> float:
        if len(self.small) == len(self.large):
            return (-self.small[0] + self.large[0]) / 2.0
        else:
            return float(self.large[0])

# Example usage
medianFinder = MedianOfStream()
medianFinder.insertNum(3)
medianFinder.insertNum(1)
print(medianFinder.findMedian())  # Output: 2.0
medianFinder.insertNum(5)
print(medianFinder.findMedian())  # Output: 3.0
medianFinder.insertNum(4)
print(medianFinder.findMedian())  # Output: 3.5